File: android/app/src/main/java/com/kvkleh/sbtsurvey/data/Formats.kt
package com.kvkleh.sbtsurvey.data

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Date, time and number formatting shared by the UI and the exporters, so that a
 * value shown on screen and the same value in the CSV always read identically.
 */
object Formats {

    private val dateFormat = SimpleDateFormat("dd-MM-yyyy", Locale.US)
    private val timeFormat = SimpleDateFormat("HH:mm:ss", Locale.US)
    private val dateTimeFormat = SimpleDateFormat("dd-MM-yyyy HH:mm", Locale.US)
    private val fileStampFormat = SimpleDateFormat("yyyyMMdd-HHmm", Locale.US)

    fun date(millis: Long?): String =
        if (millis == null) "" else dateFormat.format(Date(millis))

    fun time(millis: Long?): String =
        if (millis == null) "" else timeFormat.format(Date(millis))

    fun dateTime(millis: Long?): String =
        if (millis == null) "" else dateTimeFormat.format(Date(millis))

    fun fileStamp(millis: Long): String = fileStampFormat.format(Date(millis))

    /** Formats a coordinate with six decimals, roughly 0.1 m of resolution. */
    fun coordinate(value: Double?): String =
        if (value == null) "" else String.format(Locale.US, "%.6f", value)

    fun metres(value: Double?): String =
        if (value == null) "" else String.format(Locale.US, "%.1f", value)

    /** Renders a number without a trailing `.0`, for round values such as `2`. */
    fun number(value: Double?): String {
        if (value == null) return ""
        return if (value == Math.floor(value) && !value.isInfinite()) {
            String.format(Locale.US, "%d", value.toLong())
        } else {
            String.format(Locale.US, "%s", trimZeros(value))
        }
    }

    private fun trimZeros(value: Double): String =
        String.format(Locale.US, "%.3f", value).trimEnd('0').trimEnd('.')

    /** Parses a user-typed decimal, tolerating a comma decimal separator. */
    fun parseNumber(text: String): Double? =
        text.trim().replace(',', '.').toDoubleOrNull()

    /**
     * The Material date picker reports UTC midnight. Convert it to local midnight
     * of the same calendar day, so a harvest date picked in Leh (UTC+5:30) is not
     * stored — or exported — as the previous day.
     */
    fun pickerMillisToLocalDate(utcMillis: Long): Long {
        val utc = Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply {
            timeInMillis = utcMillis
        }
        return Calendar.getInstance().apply {
            clear()
            set(utc.get(Calendar.YEAR), utc.get(Calendar.MONTH), utc.get(Calendar.DAY_OF_MONTH))
        }.timeInMillis
    }

    /** Inverse of [pickerMillisToLocalDate], used to pre-select a stored date. */
    fun localDateToPickerMillis(localMillis: Long?): Long? {
        if (localMillis == null) return null
        val local = Calendar.getInstance().apply { timeInMillis = localMillis }
        return Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply {
            clear()
            set(
                local.get(Calendar.YEAR),
                local.get(Calendar.MONTH),
                local.get(Calendar.DAY_OF_MONTH)
            )
        }.timeInMillis
    }

    /** Local midnight for the given instant, used to store harvest dates. */
    fun startOfDay(millis: Long): Long =
        Calendar.getInstance().apply {
            timeInMillis = millis
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.timeInMillis
}

File: ios/SeabuckthornSurvey/Data/Formats.swift
import Foundation

/// Date, time and number formatting shared by the UI and the exporters, so a value
/// on screen and the same value in the CSV always read identically - and identically
/// to the Android build.
enum Formats {

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "dd-MM-yyyy"
        return formatter
    }()

    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "HH:mm:ss"
        return formatter
    }()

    private static let dateTimeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "dd-MM-yyyy HH:mm"
        return formatter
    }()

    private static let fileStampFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyyMMdd-HHmm"
        return formatter
    }()

    static func date(_ value: Date?) -> String {
        guard let value else { return "" }
        return dateFormatter.string(from: value)
    }

    static func time(_ value: Date?) -> String {
        guard let value else { return "" }
        return timeFormatter.string(from: value)
    }

    static func dateTime(_ value: Date?) -> String {
        guard let value else { return "" }
        return dateTimeFormatter.string(from: value)
    }

    static func fileStamp(_ value: Date = Date()) -> String {
        fileStampFormatter.string(from: value)
    }

    /// Six decimals, roughly 0.1 m of resolution.
    static func coordinate(_ value: Double?) -> String {
        guard let value else { return "" }
        return String(format: "%.6f", value)
    }

    static func metres(_ value: Double?) -> String {
        guard let value else { return "" }
        return String(format: "%.1f", value)
    }

    /// Renders a number without a trailing `.0`, for round values such as `2`.
    static func number(_ value: Double?) -> String {
        guard let value, value.isFinite else { return "" }
        // %lld, not %d: on a 64-bit device %d would truncate the Int64 argument.
        if value == value.rounded(.down), abs(value) < 1e15 {
            return String(format: "%lld", Int64(value))
        }
        var text = String(format: "%.3f", value)
        while text.hasSuffix("0") { text.removeLast() }
        if text.hasSuffix(".") { text.removeLast() }
        return text
    }

    /// Parses a typed decimal, tolerating a comma decimal separator.
    static func parseNumber(_ text: String) -> Double? {
        let cleaned = text.trimmingCharacters(in: .whitespaces).replacingOccurrences(of: ",", with: ".")
        guard !cleaned.isEmpty else { return nil }
        return Double(cleaned)
    }

    /// Local midnight for the given instant, used to store harvest dates.
    static func startOfDay(_ value: Date) -> Date {
        Calendar.current.startOfDay(for: value)
    }
}
